AWS Redshift has no trigger

https://docs.aws.amazon.com/ja_jp/redshift/latest/dg/c_unsupported-postgresql-features.html
↑公式ドキュメントに記載の通り、AWS Redshiftにはトリガーがありません。
そのため、直近のバージョンのtblsを実行するとpanicが発生するようです。

```
$ tbls version
dev
```

```
$ DEBUG=1 tbls doc --force --er-format svg
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x41e5af6]

goroutine 1 [running]:
database/sql.(*Rows).close(0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/opt/go/libexec/src/database/sql/sql.go:2955 +0x66
database/sql.(*Rows).Close(0x0, 0xc0002d9880, 0x2)
        /usr/local/opt/go/libexec/src/database/sql/sql.go:2951 +0x33
github.com/k1LoW/tbls/drivers/postgres.(*Postgres).Analyze(0x49d6558, 0xc0000e6180, 0xc0000a31d0, 0x468ee00, 0xc00035bc00)
        /Users/watarukura/src/github.com/k1LoW/tbls/drivers/postgres/postgres.go:158 +0x233f
github.com/k1LoW/tbls/datasource.Analyze(0xc0000b6360, 0x5c, 0x0, 0x0, 0x0)
        /Users/watarukura/src/github.com/k1LoW/tbls/datasource/datasource.go:62 +0x29d
github.com/k1LoW/tbls/cmd.glob..func2(0x49ad220, 0xc000089c20, 0x0, 0x3)
        /Users/watarukura/src/github.com/k1LoW/tbls/cmd/doc.go:72 +0x110
github.com/spf13/cobra.(*Command).execute(0x49ad220, 0xc000089bf0, 0x3, 0x3, 0x49ad220, 0xc000089bf0)
        /Users/watarukura/src/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/spf13/cobra.(*Command).ExecuteC(0x49ad940, 0xc0000f3f88, 0x4007b30, 0xc00007a058)
        /Users/watarukura/src/github.com/spf13/cobra/command.go:850 +0x2f8
github.com/spf13/cobra.(*Command).Execute(0x49ad940, 0x0, 0x0)
        /Users/watarukura/src/github.com/spf13/cobra/command.go:800 +0x2b
github.com/k1LoW/tbls/cmd.Execute()
        /Users/watarukura/src/github.com/k1LoW/tbls/cmd/root.go:59 +0x2d
main.main()
        /Users/watarukura/src/github.com/K1LoW/tbls/main.go:32 +0x20
```

/Users/watarukura/src/github.com/k1LoW/tbls/drivers/postgres/postgres.go:158

```
		// constraint
		constraintRows, err := db.Query(`
SELECT
  pc.conname AS name,
  (CASE WHEN contype='t' THEN pg_get_triggerdef((SELECT oid FROM pg_trigger WHERE tgconstraint = pc.oid LIMIT 1))
        ELSE pg_get_constraintdef(pc.oid)
   END) AS def,
  contype AS type
FROM pg_constraint AS pc
LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pc.conrelid
WHERE ps.relname = $1
AND ps.schemaname = $2
ORDER BY pc.conrelid, pc.conindid, pc.conname`, tableName, tableSchema)
		defer constraintRows.Close()
		if err != nil {
			return errors.WithStack(err)
		}
```

`SELECT oid FROM pg_trigger WHERE tgconstraint = pc.oid LIMIT 1;`

```
[42703][500310] [Amazon](500310) Invalid operation: column "tgconstraint" does not exist in pg_trigger; java.lang.RuntimeException: com.amazon.support.exceptions.ErrorException: [Amazon](500310) Invalid operation: column "tgconstraint" does not exist in pg_trigger;
```

--- a/README.md
+++ b/README.md
@@ -286,6 +286,18 @@ To set `GOOGLE_APPLICATION_CREDENTIALS` envrionment valiable, you can use
     - `?credentials=/path/to/client_secrets.json`
     - `?creds=/path/to/client_secrets.json`
 
+**Amazon Redshift:**
+
+``` yaml
+# .tbls.yml
+dsn: redshift://dbuser:dbpass@hostname:5432/dbname
+```
+
+``` yaml
+# .tbls.yml
+dsn: rs://dbuser:dbpass@hostname:5432/dbname
+```
+
 ### Document path
 
 `tbls doc` generates document in the directory specified by `docPath:`.

--- a/datasource/datasource.go
+++ b/datasource/datasource.go
@@ -13,6 +13,7 @@ import (
 	"github.com/k1LoW/tbls/drivers/bq"
 	"github.com/k1LoW/tbls/drivers/mysql"
 	"github.com/k1LoW/tbls/drivers/postgres"
+	"github.com/k1LoW/tbls/drivers/redshift"
 	"github.com/k1LoW/tbls/drivers/sqlite"
 	"github.com/k1LoW/tbls/schema"
 	"github.com/pkg/errors"
@@ -51,7 +52,11 @@ func Analyze(urlstr string) (*schema.Schema, error) {
 	switch u.Driver {
 	case "postgres":
 		s.Name = splitted[1]
-		driver = postgres.NewPostgres(db)
+		if u.Scheme == "rs" || u.Scheme == "redshift" {
+			driver = redshift.NewRedshift(db)
+		} else {
+			driver = postgres.NewPostgres(db)
+		}
 	case "mysql":
 		s.Name = splitted[1]
 		driver = mysql.NewMysql(db)

--- a/drivers/postgres/postgres.go
+++ b/drivers/postgres/postgres.go
@@ -361,7 +361,7 @@ FROM pg_constraint AS pc
 LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pc.conrelid
 WHERE ps.relname = $1
 AND ps.schemaname = $2
-ORDER BY pc.conrelid, pc.conindid, pc.conname`
+ORDER BY pc.conrelid, pc.conname`
 	}
 	return `
 SELECT

--- a/drivers/postgres/postgres.go
+++ b/drivers/postgres/postgres.go
@@ -14,14 +14,16 @@ var reFK = regexp.MustCompile(`FOREIGN KEY \((.+)\) REFERENCES ([^\s]+)\s?\((.+)
 var defaultSchemaName = "public"
 
 // Postgres struct
-type Postgres struct{
-	db *sql.DB
+type Postgres struct {
+	db     *sql.DB
+	rsMode bool
 }
 
 // NewPostgres return new Postgres
 func NewPostgres(db *sql.DB) *Postgres {
-  return &Postgres{
-		db: db,
+	return &Postgres{
+		db:     db,
+		rsMode: false,
 	}
 }
 
@@ -150,18 +152,7 @@ ORDER BY x.indexrelid
 		table.Indexes = indexes
 
 		// constraints
-		constraintRows, err := p.db.Query(`
-SELECT
-  pc.conname AS name,
-  (CASE WHEN contype='t' THEN pg_get_triggerdef((SELECT oid FROM pg_trigger WHERE tgconstraint = pc.oid LIMIT 1))
-        ELSE pg_get_constraintdef(pc.oid)
-   END) AS def,
-  contype AS type
-FROM pg_constraint AS pc
-LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pc.conrelid
-WHERE ps.relname = $1
-AND ps.schemaname = $2
-ORDER BY pc.conrelid, pc.conindid, pc.conname`, tableName, tableSchema)
+		constraintRows, err := p.db.Query(p.queryForConstraints(), tableName, tableSchema)
 		defer constraintRows.Close()
 		if err != nil {
 			return errors.WithStack(err)
@@ -195,7 +186,8 @@ ORDER BY pc.conrelid, pc.conindid, pc.conname`, tableName, tableSchema)
 		table.Constraints = constraints
 
 		// triggers
-		triggerRows, err := p.db.Query(`
+		if !p.rsMode {
+			triggerRows, err := p.db.Query(`
 SELECT tgname, pg_get_triggerdef(pt.oid)
 FROM pg_trigger AS pt
 LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pt.tgrelid
@@ -204,28 +196,29 @@ AND ps.relname = $1
 AND ps.schemaname = $2
 ORDER BY pt.tgrelid
 `, tableName, tableSchema)
-		defer triggerRows.Close()
-		if err != nil {
-			return errors.WithStack(err)
-		}
-
-		triggers := []*schema.Trigger{}
-		for triggerRows.Next() {
-			var (
-				triggerName string
-				triggerDef  string
-			)
-			err = triggerRows.Scan(&triggerName, &triggerDef)
+			defer triggerRows.Close()
 			if err != nil {
 				return errors.WithStack(err)
 			}
-			trigger := &schema.Trigger{
-				Name: triggerName,
-				Def:  triggerDef,
+
+			triggers := []*schema.Trigger{}
+			for triggerRows.Next() {
+				var (
+					triggerName string
+					triggerDef  string
+				)
+				err = triggerRows.Scan(&triggerName, &triggerDef)
+				if err != nil {
+					return errors.WithStack(err)
+				}
+				trigger := &schema.Trigger{
+					Name: triggerName,
+					Def:  triggerDef,
+				}
+				triggers = append(triggers, trigger)
 			}
-			triggers = append(triggers, trigger)
+			table.Triggers = triggers
 		}
-		table.Triggers = triggers
 
 		// columns comments
 		columnCommentRows, err := p.db.Query(`
@@ -345,6 +338,38 @@ func (p *Postgres) Info() (*schema.Driver, error) {
 		DatabaseVersion: v,
 	}
 	return d, nil
+}
+
+// EnableRsMode enable rsMode
+func (p *Postgres) EnableRsMode() {
+	p.rsMode = true
+}
+
+func (p *Postgres) queryForConstraints() string {
+	if p.rsMode {
+		return `
+SELECT
+  pc.conname AS name,
+  pg_get_constraintdef(pc.oid) AS def,
+  contype AS type
+FROM pg_constraint AS pc
+LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pc.conrelid
+WHERE ps.relname = $1
+AND ps.schemaname = $2
+ORDER BY pc.conrelid, pc.conindid, pc.conname`
+	}
+	return `
+SELECT
+  pc.conname AS name,
+  (CASE WHEN contype='t' THEN pg_get_triggerdef((SELECT oid FROM pg_trigger WHERE tgconstraint = pc.oid LIMIT 1))
+        ELSE pg_get_constraintdef(pc.oid)
+   END) AS def,
+  contype AS type
+FROM pg_constraint AS pc
+LEFT JOIN pg_stat_user_tables AS ps ON ps.relid = pc.conrelid
+WHERE ps.relname = $1
+AND ps.schemaname = $2
+ORDER BY pc.conrelid, pc.conindid, pc.conname`
 }
 
 func convertColmunType(t string, udtName string, characterMaximumLength sql.NullInt64) string {

--- a/drivers/postgres/postgres.go
+++ b/drivers/postgres/postgres.go
@@ -333,8 +333,13 @@ func (p *Postgres) Info() (*schema.Driver, error) {
 	row := p.db.QueryRow(`SELECT version();`)
 	row.Scan(&v)
 
+	name := "postgres"
+	if p.rsMode {
+		name = "redshift"
+	}
+
 	d := &schema.Driver{
-		Name:            "postgres",
+		Name:            name,
 		DatabaseVersion: v,
 	}
 	return d, nil

--- a/drivers/redshift/redshift.go
+++ b/drivers/redshift/redshift.go
@@ -1,0 +1,18 @@
+package redshift
+
+import (
+	"database/sql"
+
+	"github.com/k1LoW/tbls/drivers/postgres"
+)
+
+type Redshift struct {
+	postgres.Postgres
+}
+
+// NewRedshift return new Redshift
+func NewRedshift(db *sql.DB) *Redshift {
+	p := postgres.NewPostgres(db)
+	p.EnableRsMode()
+	return &Redshift{*p}
+}

